proposals/dom_execute.md

```diff
@@ -0,0 +1,203 @@
+# Proposal: browser.dom.execute()
+
+**Summary**
+
+This API allows an isolated world (such as a content script world or user
+script world) to synchronously execute code in the main world of the document
+its injected in.
+
+**Document Metadata**
+
+**Author:** rdcronin
+
+**Sponsoring Browser:** Chrome
+
+**Contributors:** Rob--W, ...
+
+**Created:** 2024-06-24
+
+**Related Issues:** <TODO>
+
+## Motivation
+
+### Objective
+
+This allows isolated worlds to synchronously inject in the main world.  This is
+both beneficial by itself and also a prerequisite for our current plan to
+implement inter-world communication (the latter of which will be a separate
+proposal).
+
+#### Use Cases
+
+##### Executing script in the main world
+
+Today, extensions can execute script in the main world by:
+* Leveraging scripting.executeScript() or registering content scripts or user
+  scripts, or
+* Appending a new script tag to the document
+
+The former is heavily asynchronous or requires knowledge of whether to inject
+beforehand (to register a script).  The latter is very visible to the site and
+has more potential side effects, and is also asynchronous since inline scripts
+are forbidden by the CSP applied in MV3 (to restrict remotely-hosted code
+injection).  While there's no way to fully "hide" a script from a site, it is
+desirable to avoid adding DOM elements (which could interfere with sites with
+certain properties).
+
+### Known Consumers
+
+User script managers have expressed an interest in this API, but this is also
+useful to any extension that wishes to execute script in the main world while
+leveraging information from the isolated world.
+
+## Specification
+
+### Schema
+
+```
+declare namespace dom {
+  interface ScriptInjection {
+    func: () => any;
+    args: any[];
+  };
+
+  export function execute(
+      injection: ScriptInjection
+  ): any
+}
+```
+
+### Behavior
+
+#### Function execution
+
+The function is executed _synchronously_ in the main world.  The return value
+is the serialized return value from the function execution.
+
+#### Cross-world contamination
+
+To prevent any "piercing" of the different worlds, all pieces are serialized in
+their original world, then a copy is deserialized in the target world.
+Otherwise, there would be a risk of sharing variables between worlds (even when
+non-obvious, such as by accessing the prototype of an item).
+
+##### Argument serialization
+
+Arguments are serialized and deserialized using the Structured Cloning
+algorithm.  This allows for more flexibility than simply JSON conversion.  In
+the future, some arguments may have custom serialization.
```

Comment:
If `dom.execute` doesn't support DOM nodes, it will cause problems for extension developers because dealing with DOM nodes is what content scripts usually do and `chrome.dom` namespace strongly implies that DOM is supported.

To be precise, it should be `EventTarget`, the ancestor of `Node`, which additionally allows transferring things like `iframeElem.contentWindow`, `window.parent` or `event.source` from a postMessage listener.

Although technically this is already possible by sending a `MouseEvent` with `relatedTarget` set to the EventTarget instance, but this trick is largely unknown. World isolation is maintained automatically: the event keeps only the spec'd part of the object i.e. without this JS world's expando properties.